### ISSUE TITLE
Add A 3D Level Design Tool Like Unity's Pro Builder

### ISSUE THREAD

**Issue by MonkePleb** [godotengine/godot-proposals#2879]
**Describe the project you are working on**
A 2D platformer although it's irrelevant to the feature I'm proposing.

**Describe the problem or limitation you are having in your project**
It takes game devs longer to create levels since there isn't a tool like Unity's Pro Builder in Godot.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation**
Game devs trying to create levels for a 3D game would have a much easier time with a quick and easy to use tool that's like Pro Builder.

**Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams**
I don't know as I am new to game dev. I'm currently working on my first ever game (it's a 2D game), but I noticed how there wasn't a tool like Pro Builder implemented into Godot yet.

**If this enhancement will not be used often, can it be worked around with a few lines of script?**
Probably not.

**Is there a reason why this should be core and not an add-on in the asset library?**
It should be core because it'll speed up development time for all devs working in 3D. It would especially be useful for FPS games because then levels could be created quicker. Overall, it would greatly enhance the engine and could make it a more appealing engine to new game devs who may otherwise automatically choose to use Unity.

**Comment by timothyqiu** [comment]
Maybe you can try the `CSG*` nodes for quick prototyping.

**Comment by Calinou** [comment]
> Maybe you can try the `CSG*` nodes for quick prototyping.

Unfortunately, CSG nodes have pretty bad usability when it comes to prototyping, and they're too limited to be usable in production (e.g. texturing, UV layout).

My recommendation for now is to use [Qodot](https://github.com/Shfty/qodot-plugin), but it requires using an external tool (TrenchBroom).

**Comment by gaudecker** [comment]
I agree with @Calinou that creating a full-blown level editing tool requires a lot of work, but I don't think we have to go that far (not yet anyway).

What we have now with the CSG nodes is pretty damn close. What we're missing are basic editing ergonomics, texturing options, and optimization.

#### Ergonomics

- Add gizmo for moving/resizing CSG node faces, similiar to what is described in proposal [#452](https://github.com/godotengine/godot-proposals/issues/452).
   This alone would massively improve UX.

#### Texturing

- Allow different materials for individual faces, as described in proposal [#937](https://github.com/godotengine/godot-proposals/issues/937)
- Simple tools for adjusting UV translation, rotation, and scale of each face. (also discussed in proposal [#937](https://github.com/godotengine/godot-proposals/issues/937))

#### Optimization

- A CSGCombiner3D mode that won't perform any CSG operations on its children.
   Most of my grayboxing consists of making non-overlapping boxes that do not require any CSG operations _(this is basically a Node3D with bunch of static bodies, but with added ergonomics of the CSG nodes, like easier resizing and collision)_.
- Add a "bake" button for generating single, or multiple mesh instances from a CSG tree.

-----

With these relatively simple additions alone, Godot would have a formidable toolset for prototyping levels (and even making production levels).

**Comment by Calinou** [comment]
> A CSGCombiner3D mode that won't perform any CSG operations on its children.
Most of my grayboxing consists of making non-overlapping boxes that do not require any CSG operations (this is basically a Node3D with bunch of static bodies, but with added ergonomics of the CSG nodes, like easier resizing and collision).

While this sounds interesting on paper, this might cause additional confusion in the already complex CSG node system (in my opinion).

Instead, shouldn't we enable collision generation by default on all CSG nodes? For level prototyping, you'll want most of your CSG nodes to have static collision.

**Comment by gaudecker** [comment]
> Instead, shouldn't we enable collision generation by default on all CSG nodes? For level prototyping, you'll want most of your CSG nodes to have static collision.

Much better, then you can just dump the nodes inside a Node3D and call it a day!

**Comment by Calinou** [comment]
Sometimes, I wonder if we should do away with CSGBox and CSGCylinder and just make them pre-into configured CSGPolygons (which can be regenerated at any time when desired). This would require allowing for each face to have its own material, but faces that share the same material should _not_ result in a material change to improve performance. One big upside of this approach is that you would no longer be constrained to a "perfect" box or cylinder: you could move vertices around as needed.

This could also be done for CSGSphere and CSGTorus by using the appropriate Spin modes in CSGPolygon.

**Comment by bluelightzero** [comment]
I think it woudl be better not for Godot to add a level editor, but for Godot to provide better tools for making your own level editor.

I think [EditorNode3DGizmo](https://docs.godotengine.org/en/stable/classes/class_editornode3dgizmo.html) is the first thing that could do with many improvements. It needs more options to give developers easier control over them.

Just some suggestions:

 - Allow objects to emit snapping points, which other object gizmos can snap to when they are moved near. Pass an object reference to them. Useful for allowing objects to connect to each other.
 - Allow gizmos to be selected so they can be manipulated together. Even multiple nodes and their combined gizmos.
 - Add various transformation gizmos such as rotation handles, scale handles, etc...
 - Add Node Graph handles. Nodes that when you drag, create a connection line (like node editors) rather than move the point. Useful for creating wire systems.
